[WIP] Add oc tables and details to report

sharidas opened the pull request:
Add oc table and details to the report.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

sharidas commented:
- [x] Requires core PR https://github.com/owncloud/core/pull/33845
- [x] This pr brings an addon, database schema to the config report.
- [ ] No unit test added
- [ ] Verified with command line
- [ ] Verified from the UI

sharidas commented:
- [ ] Fix the unit test failure seen in the drone.

sharidas commented:
Not sure why its failing for https://drone.owncloud.com/owncloud/configreport/285/79. Requires debugging though. But its on owncloud-daily-stable10-qa.tar.bz2.

sharidas commented:
I would like to draw attention to changes apart from code + unit test:
- [x] Had to increase the memory usage from 4096M to 8192M for `phpdbg` in Makefile. Had to debug to figure out that it was because of the memory constraint phpdbg was failing. A snapshot from IDE console is here: 
![phpstorm](https://user-images.githubusercontent.com/3600427/49988797-22e2ba80-ff9e-11e8-8558-d505983f3fdc.png) . Let me know if this looks ok.

PVince81 commented on Makefile:
```diff
@@ -13,7 +13,7 @@ occ=$(CURDIR)/../../occ
 # bin file definitions
 PHPUNIT=php -d zend.enable_gc=0  ../../lib/composer/bin/phpunit
-PHPUNITDBG=phpdbg -qrr -d memory_limit=4096M -d zend.enable_gc=0 "../../lib/composer/bin/phpunit"
+PHPUNITDBG=phpdbg -qrr -d memory_limit=8192M -d zend.enable_gc=0 "../../lib/composer/bin/phpunit"
```
@patrickjahns are we ok with this or will Drone explode eventually ?

PVince81 replied:
@sharidas I do wonder what is using all that memory. Could it be that our JSON has become too big ? But even then it shouldn't take Gigabytes

sharidas replied:
The constructor has https://github.com/owncloud/configreport/blob/master/lib/ReportDataCollector.php#L132 which is causing the problem :( I am debugging further to find the exact cause.

sharidas replied:
Machine frozen twice while debugging `listAllApps()` :(

patrickjahns replied:
Please don't raise the limit to 8GB - 4GB is more than enough.
If you need to raise the memory limit - we most probably deal with a memory leak somewhere ( recursion? ).

sharidas replied:
> Please don't raise the limit to 8GB - 4GB is more than enough.
> If you need to raise the memory limit - we most probably deal with a memory leak somewhere ( recursion? ).

Its happening in the loop https://github.com/owncloud/core/blob/master/lib/private/legacy/app.php#L781. While getting all the apps, the allocated memory 4GB is getting exhausted. For example  below is the number of bytes used in the loop ( captured with `memory_get_usage()` )
```
29001488
882905720
1734686240
1734757560
1734824432
1734824432
1734961288
2586471872
3436802864
3436802864
```

PVince81 replied:
maybe there's an infinite loop ? does the function that gather the apps also run the app's "app.php" implicitly ? are you able to get a stack trace or debug into it ?

sharidas replied:
The my investigation is making me believe that the problem is with the line https://github.com/owncloud/core/blob/master/lib/private/legacy/app.php#L812. I got the trace The trace is little large. I am using xdebug trace to grab the output to a file. The file size is little large ( ~2GB) . So need to trim it down. But yes the signature verification of the app is giving lot of function call in the trace. 

Why do I believe the problem is with the line https://github.com/owncloud/core/blob/master/lib/private/legacy/app.php#L812?
- When the line is commented and added `$result = '';` ( to confirm that I am in the right direction ), the phpunit is executed in fast manner.

sharidas replied:
Drilled down to two methods which are the memory eaters :(
- https://github.com/owncloud/core/blob/master/lib/private/IntegrityCheck/Checker.php#L342
- https://github.com/owncloud/core/blob/master/lib/private/IntegrityCheck/Checker.php#L352

Both validate signature methods are causing trouble. Debugging further.

sharidas replied:
As per my research the problem is here https://github.com/phpseclib/phpseclib/blob/2.0/phpseclib/Crypt/RSA.php#L2821-L2835

PVince81 replied:
why would that routine produce a huge stack ?

sharidas replied:
> I wonder if there are also any symlinks involved as it would likely make the integrity checker run in circles

I did not found any symlinks in the apps folder. Traversed through each apps and verified.

sharidas replied:
> As per my research the problem is here https://github.com/phpseclib/phpseclib/blob/2.0/phpseclib/Crypt/RSA.php#L2821-L2835

Again drilled down to https://github.com/phpseclib/phpseclib/blob/2.0/phpseclib/Math/BigInteger.php#L599-L602

### Remark ###
This issue is caused not when any comparison is made with the hash value in `appinfo/signature.json`. This issue is caused when the code tries to verify signatures in file
- `resources/codesigning/root.crt`
- `resources/codesigning/intermediate.crl.pem`

The hashes in the signature.json are done at the last phase of `verify()` method in the Checker.php.

PVince81 replied:
down the rabbit hole :laughing:

PVince81 commented on tests/unit/ReportDataCollectorTest.php:
```diff
@@ -92,5 +91,536 @@ public function testGetOcMigrationArray() {
+		 */
+
+		$dbtype = \OC::$server->getSystemConfig()->getValue('dbtype');
+		echo "dbtype = $dbtype\n";
```
don't echo

sharidas replied:
https://github.com/owncloud/configreport/pull/84#issuecomment-447254656

sharidas commented:
- [x] Had a closer look for the usage of `echo`. I was trying to debug and added echo and forgot to remove it. Relooked the pr, and no traces of echo are there now.

sharidas commented:
Based on my investigation/research I would like to have a summary here:

### Summary ###

- When phpdbg is ran from the command line as `phpdbg -qrr -d memory_limit=4096M -d zend.enable_gc=0 "../../lib/composer/bin/phpunit" --configuration ./phpunit.xml --testsuite unit` it has come into observation that:
   - configreports constructor calls `OC_App::listAllApps()`, which calls the IntegrityCheck/Checker.php ( here https://github.com/owncloud/core/blob/master/lib/private/IntegrityCheck/Checker.php#L577 -> https://github.com/owncloud/core/blob/master/lib/private/IntegrityCheck/Checker.php#L322)
    - The memory is issue is triggered from two calls in the `verify()` method of Checker.php:
        - https://github.com/owncloud/core/blob/master/lib/private/IntegrityCheck/Checker.php#L342
        - https://github.com/owncloud/core/blob/master/lib/private/IntegrityCheck/Checker.php#L352
    - The above mentioned `validateSignature()` call is residing in phpseclib component. And internally it calls a loop https://github.com/phpseclib/phpseclib/blob/2.0/phpseclib/Math/BigInteger.php#L599-L602. Its been noticed that around 35-40 MB of memory is consumed each time the loop is completed.
    -  Which means, each time `toBytes()` from BigInteger of phpseclib is called, around 35-40 MB of memory is used. And to our surprise its not just called at one place, there are multiple places in the code where we found the usage of `toBytes()` method.
- Another experiment I did was I tried to ran : `php -d memory_limit=4096M -d zend.enable_gc=0 "../../lib/composer/bin/phpunit" --configuration ./phpunit.xml --testsuite unit` and below is the result ( I have removed the debug outputs which I had added in the code, hence this output pasted is after removing the debugs ):
```
➜  configreport git:(add-oc-tables) ✗ php -d memory_limit=4096M -d zend.enable_gc=0 "../../lib/composer/bin/phpunit" --configuration ./phpunit.xml --testsuite unit
Cannot load Xdebug - it was already loaded
PHPUnit 5.7.27 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.2.13-1+ubuntu18.04.1+deb.sury.org+1 with Xdebug 2.6.1
Configuration: /home/sujith/test/owncloud-test/apps/configreport/phpunit.xml

..S.                                                                4 / 4 (100%)


Time: 1.08 minutes, Memory: 24.00MB

There was 1 skipped test:

1) OCA\ConfigReport\Tests\ReportDataCollectorTest::testGetOcMigrationArray
This test can be skipped for now. But needs to be fixed

/home/sujith/test/owncloud-test/apps/configreport/tests/unit/ReportDataCollectorTest.php:68

OK, but incomplete, skipped, or risky tests!
Tests: 4, Assertions: 5, Skipped: 1.

Generating code coverage report in Clover XML format ... done
```
- Another experiment was done by running 
```
➜  configreport git:(add-oc-tables) ✗ php -d memory_limit=4096M "../../lib/composer/bin/phpunit" --configuration ./phpunit.xml --testsuite unit       
Cannot load Xdebug - it was already loaded
PHPUnit 5.7.27 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.2.13-1+ubuntu18.04.1+deb.sury.org+1 with Xdebug 2.6.1
Configuration: /home/sujith/test/owncloud-test/apps/configreport/phpunit.xml

..S.                                                                4 / 4 (100%)


Time: 53.58 seconds, Memory: 24.00MB

There was 1 skipped test:

1) OCA\ConfigReport\Tests\ReportDataCollectorTest::testGetOcMigrationArray
This test can be skipped for now. But needs to be fixed

/home/sujith/test/owncloud-test/apps/configreport/tests/unit/ReportDataCollectorTest.php:68

OK, but incomplete, skipped, or risky tests!
Tests: 4, Assertions: 5, Skipped: 1.

Generating code coverage report in Clover XML format ... done
```
In the above 2 examples the difference is it was run with the `php` and not with `phpdbg`. The issue is found with `phpdbg`.

I would revert the usage of 4GB to 8GB in this PR. Also I would adjust the unit test to mock integrity check. Thanks @PVince81 for verifying my observation with phpdbg and helping hand.

PVince81 commented:
need to emphasize that the memory problem is only due to phpdbg being called with GC disabled, so whatever phpseclib does with its memory juggling will use more and more memory due to not being GC'd.

we should rather mock the integrity check service in our tests now

sharidas commented:
Current status of this PR:
- Untouched the 4GB permitted in the unit test. 
- Updated the unit test to mock the integrity check service.

PVince81 commented:
@sharidas please backport

sharidas commented:
Backport to stable10 branch https://github.com/owncloud/configreport/pull/86
